### PR TITLE
In Memory Caching

### PR DESCRIPTION
**Pull request by ardalis**
Fixes Issue #69 
@Tratcher please review

### CODE REVIEW

File: aspnet/fundamentals/caching.rst [hunk 1]
```diff
@@ -1,8 +1,157 @@
...
+	- `Caching Basics`_
+	- `Configuring In Memory Caching`_
+	- `Reading and Writing to a Memory Cache`_
+	- `Cache Dependencies and Callbacks`_
+
+`View or download sample from GitHub <https://github.com/aspnet/Docs/tree/1.0.0-beta8/aspnet/fundamentals/caching/sample>`_.
+
+Caching Basics
+--------------
+Caching can dramatically improve the performance and scalability of ASP.NET applications, by eliminating unnecessary requests to external data sources for data that hasn't changed since it was last requested. ASP.NET supports several different kinds of caches, the simplest of which is represented by the `IMemoryCache <https://github.com/aspnet/Caching/blob/dev/src/Microsoft.Extensions.Caching.Abstractions/IMemoryCache.cs>`_ interface, which represents a cache stored in the memory of the local web server.
```
Comment (Tratcher) [inline aspnet/fundamentals/caching.rst]: "for data that hasn't changed since it was last requested" -> "for data that changes less frequently than it is read"

File: aspnet/fundamentals/caching.rst [hunk 2]
```diff
@@ -1,8 +1,157 @@
...
+Configuring In Memory Caching
+-----------------------------
+To get started working with an in memory cache in your ASP.NET application, you should first add the following dependencies to your ``project.json`` file:
+
+.. literalinclude:: caching/sample/src/CachingSample/project.json
+	:linenos:
+	:lines: 5-10
+	:emphasize-lines: 4-5
+
+Caching in ASP.NET 5 is a *service* that should be referenced from your application via :doc:`dependency-injection`. To register the caching service and make it available within your application, add the following line to your ``ConfigureServices`` method in ``Startup``:
```
Comment (Tratcher) [inline aspnet/fundamentals/caching.rst]: Caching only needs to be a service if you plan to share the cache instance. A component can just as easily declare their own MemoryCache instance.

File: aspnet/fundamentals/caching.rst [hunk 3]
```diff
@@ -1,8 +1,157 @@
...
+
+There are two approaches to attempting to read entries from the cache: ``Get`` and ``TryGet``. ``Get`` will return the value if it exists, but otherwise returns ``null``. ``TryGet`` will assign the cached value to an ``out`` parameter and return true if the entry exists, otherwise it returns false.
+
+Writing to the cache is done via the ``Set`` method, which accepts the key to use to look up the value, the value to be cached, and a set of ``MemoryCacheEntryOptions``. These options allow you to specify absolute or sliding time-based cache expiration, caching priority, callbacks, and dependencies.
+
+In the ``GreetingMiddleware`` example, greetings are cached for one minute. This is implemented using the ``SetAbsoluteExpiration`` method on ``MemoryCacheOptions`` as shown here:
```
Comment (Tratcher) [inline aspnet/fundamentals/caching.rst]: MemoryCacheEntryOptions

File: aspnet/fundamentals/caching.rst [hunk 4]
```diff
@@ -1,8 +1,157 @@
...
+.. code-block:: c#
+
+	// keep item in cache as long as it is requested at least
+	// once every 5 minutes...
+	// but in any case make sure to refresh it every hour
+	new MemoryCacheEntryOptions()
+		.SetSlidingExpiration(TimeSpan.FromMinutes(5))
+		.SetAbsoluteExpiration(TimeSpan.FromHours(1))
+
+By default, an instance of ``MemoryCache`` will automatically manage the items stored, removing entries when necessary to make room for new entries. You can influence the way cache entries are managed by setting their `CacheItemPriority <https://github.com/aspnet/Caching/blob/dev/src/Microsoft.Extensions.Caching.Abstractions/CacheItemPriority.cs>`_ when adding the item to the cache. For instance, if you have an item you want to keep in the cache unless you explicitly remove it, you would use the ``NeverRemove`` priority option:
```
Comment (Tratcher) [inline aspnet/fundamentals/caching.rst]: Cache cleanup is triggered by memory pressure in the overall application, not by Sets or the size of the cache.

File: aspnet/fundamentals/caching/sample/src/CachingSample/Properties/launchSettings.json [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+      "commandName": "IISExpress",
+      "launchBrowser": true,
+      "environmentVariables": {
+        "Hosting:Environment": "Development"
+      }
+    },
+    "web": {
+      "commandName": "web",
+      "environmentVariables": {
+        "Hosting:Environment": "Development"
```
Comment (Tratcher) [inline aspnet/fundamentals/caching/sample/src/CachingSample/Properties/launchSettings.json]: We need to scrub the samples to use ASPNET_ENVIRONMENT instead of Hosting:Environment

File: aspnet/fundamentals/caching/sample/src/CachingSample/project.json [hunk 6]
```diff
@@ -0,0 +1,31 @@
+{
+  "webroot": "wwwroot",
```
Comment (Tratcher) [inline aspnet/fundamentals/caching/sample/src/CachingSample/project.json]: Remove webroot.

File: aspnet/fundamentals/caching/sample/test/CachingSample.Tests/MemoryCacheTests.cs [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+                    {
+                        _result = $"'{key}':'{value}' was evicted because: {reason}";
+                    }
+                ));
+
+            // trigger the token
+            cts.Cancel();
+
+            // wait for it
+            Thread.Sleep(100);
```
Comment (Tratcher) [inline aspnet/fundamentals/caching/sample/test/CachingSample.Tests/MemoryCacheTests.cs]: Sleeps are really bad for this, use a ManualResetEvent.

File: aspnet/fundamentals/caching/sample/test/CachingSample.Tests/project.json [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+        "System.Runtime": "4.0.21-beta-23409",
+        "System.Threading": "4.0.11-beta-23409"
+      }
+    }
+  },
+  "dependencies": {
+    "CachingSample": "1.0.0-*",
+    "xunit":"2.1.0-rc1-build3168",
+    "xunit.runner.dnx": "2.1.0-rc1-build204",
+    "Microsoft.Extensions.Caching.Abstractions": "1.0.0-rc1-final",
```
Comment (Tratcher) [inline aspnet/fundamentals/caching/sample/test/CachingSample.Tests/project.json]: No need to list transitive dependencies.

File: aspnet/fundamentals/caching/sample/test/CachingSample.Tests/project.json [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  "authors": [ "stevesmith" ],
+  "tags": [ "" ],
+  "projectUrl": "",
+  "licenseUrl": "",
+
+  "frameworks": {
+    "dnx451": { },
+    "dnxcore50": {
+      "dependencies": {
+        "Microsoft.CSharp": "4.0.1-beta-23409",
```
Comment (Tratcher) [inline aspnet/fundamentals/caching/sample/test/CachingSample.Tests/project.json]: I doubt you need any of these.

**Comment by Tratcher** [comment]
:shipit:

File: aspnet/fundamentals/caching.rst (hunk 1 above)
Comment (ardalis) [inline aspnet/fundamentals/caching.rst]: Fixed.

File: aspnet/fundamentals/caching.rst (hunk 2 above)
Comment (ardalis) [inline aspnet/fundamentals/caching.rst]: However, that will have negative coupling and testability effects (in particular if you want to test non-trivial behavior that occurs when the cache is or isn't hit), so the recommendation should be to treat it as a service, IMO.

File: aspnet/fundamentals/caching.rst (hunk 3 above)
Comment (ardalis) [inline aspnet/fundamentals/caching.rst]: Fixed

File: aspnet/fundamentals/caching.rst (hunk 4 above)
Comment (ardalis) [inline aspnet/fundamentals/caching.rst]: Clarified.

File: aspnet/fundamentals/caching/sample/src/CachingSample/Properties/launchSettings.json (hunk 5 above)
Comment (ardalis) [inline aspnet/fundamentals/caching/sample/src/CachingSample/Properties/launchSettings.json]: Is that for RC2?

File: aspnet/fundamentals/caching/sample/src/CachingSample/project.json (hunk 6 above)
Comment (ardalis) [inline aspnet/fundamentals/caching/sample/src/CachingSample/project.json]: Done.

File: aspnet/fundamentals/caching/sample/test/CachingSample.Tests/MemoryCacheTests.cs (hunk 7 above)
Comment (ardalis) [inline aspnet/fundamentals/caching/sample/test/CachingSample.Tests/MemoryCacheTests.cs]: Replaced with ManualResetEvent; tests pass; let me know if there's a better way for me to implement MRE than what I did.

File: aspnet/fundamentals/caching/sample/test/CachingSample.Tests/project.json (hunk 8 above)
Comment (ardalis) [inline aspnet/fundamentals/caching/sample/test/CachingSample.Tests/project.json]: Cleaned up.

File: aspnet/fundamentals/caching/sample/test/CachingSample.Tests/project.json (hunk 9 above)
Comment (ardalis) [inline aspnet/fundamentals/caching/sample/test/CachingSample.Tests/project.json]: Why are they in the default library template?

**Comment by ardalis** [comment]
@rachelappel should be good to merge

File: aspnet/fundamentals/distributed-cache.rst [hunk 10]
```diff
@@ -0,0 +1,10 @@
+.. _distributed-cache:
+
+.. include:: /../common/stub-topic.txt
+
+|stub-icon| Working with a Distributed Cache
```
Comment (danroth27) [inline aspnet/fundamentals/distributed-cache.rst]: @ardalis You need to include this new file in the TOC to fix the CI build error.

Comment (ardalis) [inline aspnet/fundamentals/distributed-cache.rst]: Got it, thanks.

File: aspnet/fundamentals/caching/sample/test/CachingSample.Tests/MemoryCacheTests.cs [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+        private readonly string _cacheKey = "key";
+        private string _result;
+
+        public MemoryCacheTests()
+        {
+            _memoryCache = new MemoryCache(new MemoryCacheOptions());
+        }
+
+        [Fact]
+        public void RemovalFiresCallback()
```
Comment (Tratcher) [inline aspnet/fundamentals/caching/sample/test/CachingSample.Tests/MemoryCacheTests.cs]:
```
            var pause = new ManualResetEvent(false);

            _memoryCache.Set(_cacheKey, _cacheItem,
                new MemoryCacheEntryOptions()
                .RegisterPostEvictionCallback(
                    (key, value, reason, substate) =>
                    {
                        _result = $"'{key}':'{value}' was evicted because: {reason}";
                        pause.Set();
                    }
                ));

            _memoryCache.Remove(_cacheKey);

            Assert.True(pause.WaitOne(500));

            Assert.Equal("'key':'value' was evicted because: Removed", _result);
```

File: aspnet/fundamentals/caching/sample/test/CachingSample.Tests/project.json (hunk 9 above)
Comment (Tratcher) [inline aspnet/fundamentals/caching/sample/test/CachingSample.Tests/project.json]: A stand-alone library might need them, but you pull most of them in transitively.

File: aspnet/fundamentals/caching/sample/src/CachingSample/Properties/launchSettings.json (hunk 5 above)
Comment (Tratcher) [inline aspnet/fundamentals/caching/sample/src/CachingSample/Properties/launchSettings.json]: Oh, yes.

**Comment by danroth27** [comment]
Merged 35a378c8a08e980482c6ccf73ad8ec169cb5d144

File: aspnet/fundamentals/caching/sample/test/CachingSample.Tests/MemoryCacheTests.cs (hunk 11 above)
Comment (ardalis) [inline aspnet/fundamentals/caching/sample/test/CachingSample.Tests/MemoryCacheTests.cs]: Done in a separate PR.
